### ISSUE TITLE
[Proposal] Request to add support for CI/CD pipelines for Test reporting

### ISSUE THREAD

**Issue by nigurr** [mochajs/mocha#3516]
_Issue:_ 

Currently Azure DevOps and Circle CI/CD systems provide rich test experience.
In order to get this experience/track test reporting, users has to explicitly generate test reports (currently supported format JUnit) by 3rd party plugins and publish test results.

Most of the popular open source projects aren't using 3rd party reporting plugins.
To keep local development experience and CI/CD build experience similar, scripts to build/test are written to print test results to console (std_out).
To get richer test experience in CI/CD pipeline, customers has to modify their scripts and often need to add 3rd party plugins.
This is a steep learning and test reporting is one of the important data that should be easily available for open source github projects.

_Proposal for solution(s):_

**Generate test results file based on environment variables**:
     In this approach, test runners will generate test results file along with regular console output if an environment variable is set.
     CI/CD pipelines can set this variable to desired test output location where they can automatically publish test results by consuming test results file generated by test runners (mocha).
     This simple flag will make sure test runners are not going through any change in behavior in terms of execution or context they are running and users don't require go through any modification to their source/scripts.

Happy to listen to community feedback. Feel free to share other ideas if proposed solution is not feasible one.

**Comment by nigurr** [comment]
I would be very happy to answer any queries, will be looking forward to listen to community feedback.

**Comment by plroebuck** [comment]
> users has to explicitly generate test reports (currently supported format JUnit)

JUnit is a test specification format for Java, not a report.
And CircleCI does **not** specify a report format.

> Most of the popular open source projects aren't using 3rd party reporting plugins.

Doesn't matter whether most open source projects are using 3rd party reporters or not; if they use Mocha, you can install your own reporter and use that instead. But not sure why you're trying to run tests on open source projects rather than your own code.

> To get richer test experience in CI/CD pipeline, customers has to modify their scripts and often need to add 3rd party plugins.

And? Par for the course, no?

> In this approach, test runners will generate test results file along with regular console output if an environment variable is set. CI/CD pipelines can set this variable to desired test output location where they can automatically publish test results by consuming test results file generated by test runners (mocha).

You mean a CircleCI setup that will write to both screen and output file?
```yaml
steps:
- run:
    environment:
      REPORTER: "json"
      REPORT_DIR: "/path/to/reports/dir"
    mkdir -p ${REPORT_DIR}
    npm test -- --reporter ${REPORTER} | tee ${REPORT_DIR}/test-${REPORTER}.out
```

Basically, it's on _you_ to setup your CI/CD environment.

**Comment by plroebuck** [comment]
> > > users has to explicitly generate test reports (currently supported format JUnit)
> > 
> > JUnit is a test specification format for Java, not a report.
> > And CircleCI does **not** specify a report format.
> 
> Sorry if it's not clear. To get rich test results experience in Circle CI/Azure DevOps pipeline, currently we need to publish test artifacts as Junit format
> ref: https://circleci.com/docs/2.0/collect-test-data/ https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/test/publish-test-results?view=vsts&tabs=yaml
> 
> Junit is a test specification format and it's being used as standard in multiple runners to report results (PyTest, PHPUnit)
> Proposal is to generate test results, it could be _mocha json format/Junit xml_ or anything.

So as I understand what you're in reference to now, CircleCI can provide additonal rich test results if the test output is formatted in JUnit's XML output format. OK. But you can't overgeneralize here, since (to my knowledge) most other CIs (e.g., Travis, Appveyor) don't -- it's _specific_ to your CI pipeline.

> > > Most of the popular open source projects aren't using 3rd party reporting plugins.
> > 
> > Doesn't matter whether most open source projects are using 3rd party reporters or not; if they use Mocha, you can install your own reporter and use that instead. But not sure why you're trying to run tests on open source projects rather than your own code.
> 
> Agree on this. However to keep local development/pipeline experience same, users are not adding extra reporter for CI Pipeline alone.
> Github analytics shows that only 25% of mocha users having custom reporter and 75% of users are using standard console output.
> 
> This leaves pipelines not to get any test information as make script will print information to standard console by default.
> TravisCI currently doesn't have any experience around this. However CircleCI and Azure Devops both supports rich test experience along with test analytics.

Since most CIs do nothing special based on output format type, it shouldn't be surprising that many users don't bother to install a custom reporter.

> > > To get richer test experience in CI/CD pipeline, customers has to modify their scripts and often need to add 3rd party plugins.
> > 
> > Par for the course, no?
> 
> It's a steep learning for customers to understand test reporting capabilities and easily diagnose the test information.
> As an example, currently customers has to go through build log to see where the test has failed and get to stack trace.
> With test results experience, customers can directly navigate to Tests hub and get rich experience of the same.

Not following where "customers" entered the picture. So customers are not the developers on your project? Or do you work for Azure DevOps?

> > > In this approach, test runners will generate test results file along with regular console output if an environment variable is set. CI/CD pipelines can set this variable to desired test output location where they can automatically publish test results by consuming test results file generated by test runners (mocha).
> > 
> > You mean a CircleCI setup that will write to both screen and output file? No change required.
> > ```yaml
> > steps:
> > - run:
> >     environment:
> >       REPORTER: "json"
> >       REPORT_DIR: "/path/to/reports/dir"
> >     mkdir -p ${REPORT_DIR}
> >     npm test -- --reporter ${REPORTER} | tee ${REPORT_DIR}/test-${REPORTER}.out
> > ```
> > 
> > Basically, it's on _you_ to setup your CI/CD environment.
> 
> Agree on this if users have clear differentiation between each step and explored pipeline features.
> This is steep learning curve and users can get benefited out of the box if pipelines can provide these features out of the box.

So now you reference users. Are they customers? developers? other (e.g., mgmt)?
So again, regardless, if you're worried about the rich test experience, set it up yourself.
We cannot default to the CircleCI rich reporting experience, as Mocha itself doesn't provide a JUnit reporter.

> As mentioned in CircleCi yaml, one has to currently add [Junit reporter](https://www.npmjs.com/package/mocha-junit-reporter), [Multi reporter](https://www.npmjs.com/package/mocha-multi-reporters) to generate JUnit test reports along with console output. Once it's added to source code, one has to explicitly `publish test artifacts` to get richer experience.

So add the steps to your CircleCI YAML file (untested).
```yaml
steps:
  - run:
      name: Install 3rd party Mocha reporters
      command: |
        npm install mocha-multi-reporters mocha-junit-reporter --save-dev
  - run:
      name: Setup shared/interpolated environment variables
      command: |
        echo 'export PROPERTIES="${BUILD_ID}:${CIRCLE_BUILD_NUM}"' >> ${BASH_ENV}
        echo 'export CONFIG_FILE="multi-reporters-config.json"' >> ${BASH_ENV}
        echo 'export REPORT_DIR="/path/to/reports/dir"' >> ${BASH_ENV}
        echo 'export MOCHA_FILE="${REPORT_DIR}/mocha-junit.xml"' >> ${BASH_ENV}
        source ${BASH_ENV}
  - run:
      name: Write simple MultiReporters config file, enabling `spec` and `JUnit` reporters
      command: |
        cat '{ "reporterEnabled": "spec, mocha-junit-reporter" } > ${CONFIG_FILE}
  - run:
      name: Create reports directory, if necessary
      command: |
        mkdir -p ${REPORT_DIR}
  - run:
      name: Run tests
      environment:
         REPORTER: "mocha-multi-reporters"
      command: |
        npm test -- --reporter ${REPORTER} --reporter-options configFile=${CONFIG_FILE}
  - store_artifacts:
      ## Don't think envvars get evaluated here. Repeat value of `REPORT_DIR` from above
      path: /path/to/reports/dir
```


> **Proposal**
> 
>     1. Mocha accepts environment variable to support custom arguments. Similar to [Pytest AddOps](https://docs.pytest.org/en/latest/customize.html)
>        Example: --reporter json

We have cmdline args, as well as "mocha.opts". Not seeing adding this "feature" without significantly more upvotes. You technically already have this capability via shell variables.
```bash
$ MOCHA_OPTS="--reporter json"
$ mocha ${MOCHA_OPTS}
```

>     2. Pipelines by default add this environment variable so that mocha runners will generate test results to json out of the box

Uh, doubtful. This would not be up to us (or you).

>     3. Pipelines will automatically detect and publish the test results to lighten the [test experience](https://dev.azure.com/dtatestingint/PBoraProjects/_build/results?buildId=6623&view=ms.vss-test-web.test-result-details).

No, they wouldn't - you realize those test artifacts take disk space as well as time to move around, as well as additional time/cpu to process after reimport. That's why it's opt-in processing.

> **Conclusion**
> If Mocha supports an environment variable `(MOCHA_OPTS)` which can be used to pass additional arguments, CI providers can generate test results and automatically publish out of the box.

Your conclusion makes _way_ too many assumptions.

**Comment by rishavsharan** [comment]
Hi @plroebuck 
Myself and @nigurr work on the Azure DevOps product. Our customers/users are the organizations and developers using our Pipelines to Build and Release their products.

> So as I understand what you're in reference to now, CircleCI can provide additonal rich test results if the test output is formatted in JUnit's XML output format. OK. But you can't overgeneralize here, since (to my knowledge) most other CIs (e.g., Travis, Appveyor) don't -- it's specific to your CI pipeline.

Most Ci/CD platforms currently support jUnit-esque format to show rich test metadata.
For example;
Appveyor - https://www.appveyor.com/docs/running-tests/#uploading-xml-test-results
GitLab - https://docs.gitlab.com/ee/ci/junit_test_reports.html
Bitbucket - https://confluence.atlassian.com/bitbucket/test-reporting-in-pipelines-939708543.html
CircleCi- https://circleci.com/docs/2.0/collect-test-data/
etc.
Test Reporting in Travis is very rudimentary right now and, as a result, it doesn't needs the xml.

Its currently a common format amongst all these platforms and having mocha output the result xml will benefit all players in the ecosystem.

**Comment by plroebuck** [comment]
> Hi @plroebuck
> Myself and @nigurr work on the Azure DevOps product. Our customers/users are the organizations and developers using our Pipelines to Build and Release their products.

Hello, @rishavsharan. Thanks for identifying yourself and your interest.

> > So as I understand what you're in reference to now, CircleCI can provide additonal rich test results if the test output is formatted in JUnit's XML output format. OK. But you can't overgeneralize here, since (to my knowledge) most other CIs (e.g., Travis, Appveyor) don't -- it's specific to your CI pipeline.
> 
> Most Ci/CD platforms currently support jUnit-esque format to show rich test metadata.
> For example;
> Appveyor - https://www.appveyor.com/docs/running-tests/#uploading-xml-test-results
> GitLab - https://docs.gitlab.com/ee/ci/junit_test_reports.html
> Bitbucket - https://confluence.atlassian.com/bitbucket/test-reporting-in-pipelines-939708543.html
> CircleCi- https://circleci.com/docs/2.0/collect-test-data/
> etc.
> Test Reporting in Travis is very rudimentary right now and, as a result, it doesn't needs the xml.
> 
> Its currently a common format amongst all these platforms and having mocha output the result xml will benefit all players in the ecosystem.

Huh, never really bothered to look for such a capability, treating the output as success/failure.
But since you pointed out [AppVeyor](https://www.appveyor.com/docs/running-tests/#uploading-xml-test-results) having this capability, I note it does support xUnit. Was unaware of GitLab's newfound capability. Bitbucket also allows xUnit. Reading further on [CircleCI](https://circleci.com/docs/2.0/collect-test-data/#ava-for-nodejs), would seem they support xUnit as well (though not explicitly spelled out).

Since Mocha supports [xUnit](https://mochajs.org/#xunit) out-of-the-box, does that reporter not work for your pipelines?

**Comment by rishavsharan** [comment]
@plroebuck Thanks for the link to the xunit reporter. This will be very helpful for our usecase. 

Regarding the 2nd request of having environment variables to add/override the mocha configs, I fear shell variables will not work for us. To be able to use shell variables, the platform will have to know that the current project is actually running mocha as the test framework. Environment variables on the other hand are passively defined and the platform will not need to know anything about the project details. If it runs mocha, the config will automatically be picked up.
This will allow us to do very platform specific things like run custom config that we need to generate test metadata and run custom reporters. 

This feature will help all CI/CD platforms, and not just Azure DevOps, specially as they begin to focus more on the test reporting and analysis part and will ultimately help mocha users leverage such platforms like Azure DevOps, without writing any platform specific test config.

**Comment by plroebuck** [comment]
> Regarding the 2nd request of having environment variables to add/override the mocha configs,
> I fear shell variables will not work for us. To be able to use shell variables, the platform will have
> to know that the current project is actually running mocha as the test framework. Environment 
> variables on the other hand are passively consumed and the platform will not need to know 
> anything about the project details. If it runs mocha, the config will automatically be picked up.

Sorry, but my reading of what you just said would imply one of two things:
1. You're going to create a `ONE_TO_RULE_THEM_ALL` envvar for _all_ test frameworks/runners.
1. You plan to create a fleet of envvars settings, pass them to pipeline where the project's test runner will pick it up the appropriate one automagically. (more likely)

There are platform-specific [limits](https://askubuntu.com/questions/14081/what-is-the-maximum-length-of-command-line-arguments-in-gnome-terminal) and the latter implication will potentially cause other issues.
Hint: Limits for Windows are _far_ smaller than those for Linux.

> This will allow us to do very platform specific things like run custom config that we need to 
> generate test metadata and run custom reporters even if all the user has defined "npm test"
> in their pipelines task.

Sure you're aware that cmdline args _trump_ envvars. So your idea would not work unless the project was relying on default settings. You could also potentially break project workflows by introducing such changes.

> This feature will help all CI/CD platforms, and not just Azure DevOps, specially as they begin to 
> focus more on the test reporting and analysis part and will ultimately help mocha users leverage 
> such platforms like Azure DevOps, without writing any platform specific test config.

Just me -- I disagree with this concept entirely. If I want something, I can ask. If not, I would expect **no** interference from a pipeline. What you're talking about doing _must_ be done "opt-in".

> Finally, if the maintainers are fine with the feature request, we would love to contribute to the project and submit PRs for this.

We welcome PRs. No guarantees on acceptance though.
